src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,369 @@
+            Debug.Assert(source.Length <= destination.Length);
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                char c = source[i];
+                if (char.IsHighSurrogate(c) && i < source.Length - 1)
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToUpper(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToUpper(c);
+            }
+        }
+
+        internal static void ToLower(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+            Debug.Assert(source.Length <= destination.Length);
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                char c = source[i];
+                if (char.IsHighSurrogate(c) && i < source.Length - 1 )
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToLower(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToLower(c);
+            }
+        }
+
+        internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char strB, int lengthB)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+
+            int length = Math.Min(lengthA, lengthB);
+
+            ref char charA = ref strA;
+            ref char charB = ref strB;
+
+            while (length != 0)
+            {
+                if (!char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                {
+                    if (charA == charB)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    char aUpper = ToUpper(charA);
+                    char bUpper = ToUpper(charB);
+
+                    if (aUpper == bUpper)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    return aUpper - bUpper;
```

Comment:
I think there's a bug here. Say that I have these two input strings, each exactly 7 chars long:

_strA_: `"Hello\uFE6A!"` (where `U+FE6A` = `'﹪' (SMALL PERCENT SIGN)`)
_strB_: `"Hello\U0001F601"` (where `U+1F601` = `"😁" (GRINNING FACE WITH SMILING EYES)`)

_strA_ should ostensibly sort __before__ _strB_ under _OrdinalIgnoreCase_ sorting, as neither the `U+FE6A` code point nor the `U+1F601` code point changes case, and `0xFE6A < 0x1F601`.

However, per the implementation of this logic, when we get to _index = 5_ in each string, we'll compare `'\uFE6A'` against `'\uD83D'` (the high surrogate char), see that neither changes case, then determine that `0xFE6A > 0xD83D` and sort _strB_ __before__ _strA_.

Is _strB_ sorting before _strA_ under an ordinal comparer intentional? I find it somewhat surprising.

Reply:
Good catch. I have re-written the method in a clean way and should be handling surrogate correctly. 

Note, I am seeing same problem in another place in the OrdinalCasing. I'll address this part in a different PR to look at the perf when changing it there.